### PR TITLE
kubeadm: wait for the etcd cluster to be available when growing it

### LINKED ISSUE #1353: `kubeadm join` does not explicitly wait for etcd to have grown when joining secondary control plane
## Is this a BUG REPORT or FEATURE REQUEST?

BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`): `1.13` onwards.

## What happened?

When joining a secondary control plane (`kubeadm join --experimental-control-plane`, or by using `controlPlane` configuration on the `kubeadm` YAML), and `etcd` needs growing from 1 instance to 2 there's a temporary blackout that we are not explicitly waiting for.

This is a spin-off of https://github.com/kubernetes/kubeadm/issues/1321#issuecomment-454384232.

Different issues were leading to https://github.com/kubernetes/kubeadm/issues/1321 error on the new joining master: `error uploading configuration: Get https://10.11.0.1:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config: unexpected EOF`.

1) https://github.com/kubernetes/kubeadm/issues/1321
    * Bad certificates when growing etcd. More certificates than the required ones are copied between master machines; `kubeadm` won't recreate them (as they exist), but their SANs are wrong and `etcd` will fail to grow to 2 nodes. The previous error happens because is the immediate next step after we write the etcd static pod manifests on the new master.

2) https://github.com/kubernetes/kubeadm/issues/1341
    * When joining a new controller plane we are not pre-pulling the images
    * The `uploadconfig` phase can timeout with the previous mentioned error because the system is still pulling images while we are trying to upload the new cluster configuration (with the new node)

3) This issue
    * When we grow `etcd` from 1 instance to 2 there's a blackout of etcd that we should explicitly wait for. Instead, we are at the moment implicitly waiting for it in this phase (`uploadconfig`). If this blackout is long enough (many index sync required, slow I/O), the `uploadconfig` phase can timeout with the error mentioned before.

## What you expected to happen?

When we grow `etcd` we should perform an intermediate step explicitly waiting for `etcd` to be available (or healthy). Then, we can continue with the next step, in this case `uploadconfig`.

The timeout happens here because `uploadconfig` happens to be the first phase after the static etcd pod creation that tries to reach to the apiserver while the etcd blackout is still happening, when it tries to call to `GetClusterStatus`.

## How to reproduce it (as minimally and precisely as possible)?

Please, refer to https://github.com/kubernetes/kubeadm/issues/1321#issuecomment-454384232

### PR DESCRIPTION
**Pull request by ereslibre**
**What this PR does / why we need it**:

When the etcd cluster grows we need to explicitly wait for it to be
available. This ensures that we are not implicitly doing this in
following steps when they try to access the apiserver.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubeadm/issues/1353

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: explicitly wait for `etcd` to have grown when joining a new control plane
```

/kind bug

### CODE REVIEW

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (ereslibre) [inline cmd/kubeadm/app/util/etcd/etcd.go]: This is for making `AddMember` automatically modify the client endpoints without having to reach anything externally (etcd or the apiserver). However I don't like this because it's not really needed to pass the client address when adding a member. Also, feels hackish to compute this inside the `AddMember` function because the peer address is a complete URI (with scheme and port), and this function could compute it but would have to do tricky operations with it to set the appropriate port and we don't have the init config here to call to `GetClientURL`.

Honestly, if this is too bad maybe we can go back to the alternative option where `AddMember` does not change the client and we manually add the endpoint on `CheckLocalEtcdClusterStatus`.

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/etcd/local.go]: what do your tests reveal for the interval and n-retries @ereslibre ?

i think the 1 second rate might be too high. i would do something like 5 seconds, with 20 retries.
but let's gather more comments on this one before changing.

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd.go]: as a note/TODO: in a separate PR  we need to make all the client methods to use pointers.
golang encourages a matching pattern.

**Review by neolit123 — APPROVED** [review]
thanks for this @ereslibre 
i think apart from the retry rate this is good.

also please add a release note instead of `NONE`
`kubeadm: ......`

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/etcd/local.go]: missed this one, please use `fmt.Println` as there are no args.

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (rosti) [inline cmd/kubeadm/app/util/etcd/etcd.go]: A few things here:
1. You are parsing the peer address, while the endpoints point to the client address. In most cases they differ only in port number, but let's just be safe here...
2. The call to `url.Parse` (in theory) can fail after successful call to `MemberAdd`, thus failing the `AddMember` while we actually have a new member.
3. You don't need to call `url.Parse` and `GetClientURLByIP`. The client address for the new node is in `resp.Member.ClientURLs`.
4. All of the active members (as of the call to `MemberAdd`) should be returned in `resp.Members`. You can iterate them and get their `ClientURLs` to regenerate the latest version of `c.Endpoints`.

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (rosti) [inline cmd/kubeadm/app/util/etcd/etcd.go]: This change has the potential to spew a lot of endpoints in a regular message here. I would recommend moving the endpoint dump to a separate `klog.V(2).Infof(...)` message right after this one.

**Review by rosti — CHANGES_REQUESTED** [review]
Thanks @ereslibre !
Overall it looks good, though we need to fix the `AddMember` call.

**Comment by rosti** [comment]
@ereslibre we need to have a release note as there is a released version of kubeadm with this bug and we need to indicate a merged fix for it.
We'll also need to cherry-pick this in 1.13 once it merges.

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/etcd/etcd.go]: Let me explain the process I went through, and why I think this is the less bad solution:

1. on the local etcd setup we [control the endpoints](https://github.com/kubernetes/kubernetes/blob/af31a202ea5ac80ebaca81fa949bf3b31c649618/cmd/kubeadm/app/phases/etcd/local.go#L157) and it's basically going to be this hostname with this port.

2. I didn't think of this case, is this feasible? If the provided peer address is not an URL, how can the member addition succeed?

3. This was my very first try, but the ClientURLs is empty until the `etcd` member starts, and at this point in the code the kubelet is basically starting it. On the response we only have the peer address with an empty list of client URLS, so this is not really an option :(

4. This is the same issue, the new added member is not active and hasn't yet got ClientURLs; it could have (depending if etcd actually started on the background or not).

So I went through the process of checking what options we did have and I tried:

a) Call to `CheckLocalEtcdStatus` directly: the straightforward one. Since this tries to retrieve the endpoints from `GetClusterStatus` on `NewFromCluster` is again a suboptimal situation, because we are blocking trying to download the `kubeadm-config` configmap during the blackout.

b) Call to `Sync()` on etcdclient: during the blackout this again not the best idea, because we would have to loop until we get a valid answer, felt awkward.

c) Try to get the real response from `MemberAdd`: empty clientURLs because the instance could have not started yet; only the peerURLs is populated for the new member.

d) Add the endpoint manually, parsing the hostname and changing the port to the client port.

I also have thought if it's worth checking if this endpoint is on the list to avoid adding duplicates (it wouldn't be a problem, we would check the new endpoint more than once if it was already there, but I was wondering if the hassle of checking if the endpoint exists to avoid checking it twice is worth it).

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local.go]: In my environment it's succeeding around the third time. An environment that has been alive for seconds, so I agree that 5 seconds looks reasonable (if the cluster of one machine is long-living more sync would be neded).

I'll adapt the PR, thanks!

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (rosti) [reply cmd/kubeadm/app/util/etcd/etcd.go]: Thanks for the elaboration @ereslibre ! Since you have verified that this is the best working solution, then I am OK to keep it that way.

Comment (ereslibre) [reply cmd/kubeadm/app/util/etcd/etcd.go]: Thank you @rosti for the feedback!

**Comment by rosti** [comment]
@ereslibre Just a tiny nit on the release note. Can you prefix it with `kubeadm: ` so that it's clear, that this goes into the kubeadm change set?

**Review by rosti — APPROVED** [review]
Re-lgtm after things are back to normal.
/lgtm

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/etcd/local.go]: If you don't mind, could you change "local etcd instance" into "local etcd member" (so we are consistent in using always etcd member)

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/etcd/local.go]: I don't like this function printing
```
    [util/etcd] Attempt timed out
    [util/etcd] Waiting 5s until next retry
    [util/etcd] Attempt timed out
    [util/etcd] Waiting 5s until next retry
```
IMO those output should be removed (or converted into log messages) in order to be consistent with all the other waiters in kubeadm.

However, considering that this requires to add "This can take up to ..." in every place where `WaitForClusterAvailable` is used, this goes out of the scope of this PR, so please open an issue to track this as a todo/good first issue

File: cmd/kubeadm/app/util/etcd/etcd.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/util/etcd/etcd.go]: If we move only the `url.Parse` + `if err != nil`  at the beginning of the function we can address @rosti comment "The call to url.Parse (in theory) can fail after successful call to MemberAdd, thus failing the AddMember while we actually have a new member." while preserving your working solution.
wdyt?

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@ereslibre Great work! this is a candidate a cherry pick!
Only a few small nits mostly UX related
Ping me when this is ready for approval

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local.go]: I proposed to use `klog` here too but @rosti didn't want to address this change on this PR, only changing the one potentially long with the endpoints. I agree with your point of view though @fabriziopandini.

@rosti, wdyt? Should I change this now that @fabriziopandini also raised this issue?

Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local.go]: Marked as resolved as per discussion with @fabriziopandini, leaving it as it was as @rosti proposed.

Comment (rosti) [reply cmd/kubeadm/app/phases/etcd/local.go]: I do think, that we need some sort of indication about the reason we wait another 5 seconds. This is tightly coupled with the UX of end users, that run kubeadm directly on command line. For that matter I am not a fan of klogging this. In my opinion it should go out via print.
On the other hand, we can certainly reduce the output here to say a single, more descriptive message per retry.
However, as @fabriziopandini mentioned, this will require changes in a few more places, thus it may better be done in another PR. We can file a backlog issue for now.

**Comment by rosti** [comment]
@ereslibre can you, please, re-run the `update-gofmt.sh` script to fix the verify test? Thanks!

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/etcd/local.go]: i wanted to get more feedback on the 5 seconds and the interval of 20 tries.
if we can get a check faster than 5 on the average, possibly we can reduce the value?
also 20 tries is a lot. in reality, we might get the failed state much sooner.

Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local.go]: On my environment that is fresh everytime it took 3 retries when it was a 1 second interval, and right now with 5 seconds it retries none or once.

Without feedback on how much this takes on different environments it's hard to know (I guess it depends on the uptime of the cluster and the I/O throughput). That being said, my concern is that if we lower the timeout too much we might call it a failure when a long living one-master cluster (weeks, months?) gets a new master.

Just my 2 cents, but of course it would be great to have real data on this.

Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local.go]: In the last run it took 4 retries (of 5 seconds interval), this one was way off-charts and with a clean environment :(

Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local.go]: I also find a bit weird the output of the `kubelet-check` in the middle TBH:

```
[etcd] Waiting for the new etcd member to join the cluster. This can take up to 1m40s
I0118 12:22:43.405682    2420 etcd.go:270] attempting to see if all cluster endpoints ([https://172.28.128.56:2379 https://172.28.128.57:2379]) are available 1/20
[util/etcd] Attempt timed out
[util/etcd] Waiting 5s until next retry
I0118 12:22:53.502975    2420 etcd.go:270] attempting to see if all cluster endpoints ([https://172.28.128.56:2379 https://172.28.128.57:2379]) are available 2/20
[util/etcd] Attempt timed out
[util/etcd] Waiting 5s until next retry
I0118 12:23:03.590413    2420 etcd.go:270] attempting to see if all cluster endpoints ([https://172.28.128.56:2379 https://172.28.128.57:2379]) are available 3/20
[util/etcd] Attempt timed out
[util/etcd] Waiting 5s until next retry
[kubelet-check] Initial timeout of 40s passed.
I0118 12:23:13.615829    2420 etcd.go:270] attempting to see if all cluster endpoints ([https://172.28.128.56:2379 https://172.28.128.57:2379]) are available 4/20
I0118 12:23:13.657347    2420 join.go:625] [join] uploading currently used configuration to the cluster
```

coming from https://github.com/kubernetes/kubernetes/blob/26d58d37a71ecc698259729361484007b45a82ba/cmd/kubeadm/app/util/apiclient/wait.go#L163-L167

Comment (neolit123) [reply cmd/kubeadm/app/phases/etcd/local.go]: i've mentioned this on slack:
> we may want to keep the overall time under 40seconds to match the kubelet timeout.
how about 2 seconds rate with 20 retries.

Comment (neolit123) [reply cmd/kubeadm/app/phases/etcd/local.go]: @fabriziopandini @rosti please give you stamp of approval for the above comment.

Comment (rosti) [reply cmd/kubeadm/app/phases/etcd/local.go]: I like the 40 seconds idea, but let's keep the steps at 5 sec. Bear in mind, that we have just written out the static pod, so the kubelet needs to detect it, spin it up and for etcd to become responsive. On some systems it's easy for this to come above 2 seconds.

Comment (neolit123) [reply cmd/kubeadm/app/phases/etcd/local.go]: ok, @rosti is voting for 5 sec / 8 retries.
@fabriziopandini ?

**Comment by fabriziopandini** [comment]
Looking at the code I don't see a relation between the kubelet timeout for TLS bottstrap and the timeout this PR sets for waiting for the new etcd member joining the cluster.

Said that, IMO  5s * 8 is reasonable for unblocking this fix and start the cherry-picking process; 
so, considering that it seems there is consensus on those settings also from the other reviewers, and that the proposed solution is definetly an improvement vs current situation
/approve
/lgtm

Let's continue the discussion in parallel (slack/at the next office hours meeting) eventually asking to broader sig for some more feedbacks from the field

**Comment by MalloZup** [comment]
We could also use kind of backoff algo for don't use a fixed timeout.  There are also golang lib out there https://en.m.wikipedia.org/wiki/Exponential_backoff?wprov=sfla1

**Comment by ereslibre** [comment]
@fabriziopandini I also have the same feeling and have discussed it with @neolit123 previously. So, just for the record, `WaitForKubeletAndFunc` basically waits for two things:

1) `WaitForHealthyKubelet` before some initial timeout (40 seconds)
2) The given function to `WaitForKubeletAndFunc`

When any of these two functions (given one or `WaitForHealthyKubelet`) returns, `WaitForKubeletAndFunc` returns. If the given function takes more than 40 seconds to finish, then the sleep on `WaitForHealthyKubelet` timeouts and *actually* checks that the kubelet is healthy.

I guess that I would say `WaitForKubeletAndFunc` basically means "please, run this function I give you, if it fails or succeeds it's fine, *but* if it takes more than 40 seconds to answer, then we check that the kubelet is running and healthy"

**Pull request merged**
